WIP SOC8 support

bmwiedemann opened the pull request.

bmwiedemann committed:
Update to SLE-12-SP3

SES5 only supports this

bmwiedemann committed:
ceph: deploy ceph-mgr

Deploy it after the admin keyring creation.

rsalevsky commented:
Please note master is in this case Cloud 8 and we actually disabled support for it.

tserong commented on chef/cookbooks/ceph/recipes/mon.rb:
```diff
@@ -30,9 +30,50 @@
+unless File.exist?("/var/lib/ceph/mgr/ceph-#{mon_name}/done")
+  keyring = "#{Chef::Config[:file_cache_path]}/#{cluster}-#{mon_name}.mgr.keyring"
+  execute "create mgr keyring" do
+    command "ceph-authtool '#{keyring}' --create-keyring --name=mgr. --add-key='#{node["ceph"]["monitor-secret"]}' --cap mgr 'allow *'"
```
The mgr keyring needs to be written to `/var/lib/ceph/mgr/ceph-#{mon_name}/keyring` and needs caps `mon 'allow profile mgr' osd 'allow *' mds 'allow *'`, something like the "create mds keyring" block in chef/cookbooks/ceph/recipes/mds.rb.

tserong reviewed (changes requested):
Aside from the keyring change, this looks like it will work to deploy ceph-mgr.

Note though that SES5 doesn't include Calamari, so if SOC8 ends up supporting deploying SES5, the Calamari role will need to be removed. This leaves the Ceph cluster without a GUI, unless we were to add an openATTIC role, to deploy openATTIC instead, but that gets more complicated, as parts of openATTIC rely on DeepSea, which won't be present if SES is deployed via Crowbar.

bmwiedemann commented:
Added that, but somehow there is no admin keyring.
Maybe something else needs to be done first.
Addding `include_recipe "ceph::keyring"` also does not help

```
ceph auth get-or-create mgr.public.d52-54-77-77-01-03              mon 'allow profile mgr' osd 'allow *' mds 'allow *'              -o /var/lib/ceph/mgr/ceph-public.d52-54-77-77-01-03/keyring
2018-10-01 13:44:28.538622 7f52022dc700 -1 auth: unable to find a keyring on /etc/ceph/ceph.client.admin.keyring,/etc/ceph/ceph.keyring,/etc/ceph/keyring,/etc/ceph/keyring.bin,: (2) No such file or directory
2018-10-01 13:44:28.538650 7f52022dc700 -1 monclient: ERROR: missing keyring, cannot use cephx for authentication
2018-10-01 13:44:28.538653 7f52022dc700  0 librados: client.admin initialization error (2) No such file or directory
[errno 2] error connecting to the cluster
```

tserong commented:
Try moving all the mgr creation stuff down to the end of the file; ceph.client.admin.keyring isn't created yet until about line 200.

bmwiedemann committed:
Adapt radosgw to keystone v3

Needed after crowbar-openstack commit 8774f1a509
Use keystone v3 in config auth urls (SCRD-781)

bmwiedemann commented:
got a bit further now. I was able to adapt radosgw recipes to keystonev3
I never used a ceph UI so far, so I guess, dropping calamares can be OK for now.

```
Mixlib::ShellOut::ShellCommandFailed
------------------------------------
Expected process to exit with [0], but received '1'
---- Begin output of chown ceph:ceph /var/chef/cache/ceph-public.d52-54-77-77-01-01.mon.keyring ; ceph-mon --mkfs -i public.d52-54-77-77-01-01 --keyring '/var/chef/cache/ceph-public.d52-54-77-77-01-01.mon.keyring' --setuser ceph --setgroup ceph ----
STDOUT: 
STDERR: chown: cannot access '/var/chef/cache/ceph-public.d52-54-77-77-01-01.mon.keyring': No such file or directory
2018-10-04 12:24:53.035565 7f0168094f40 -1 mon.public.d52-54-77-77-01-01@-1(probing) e0 unable to find a keyring file on /var/chef/cache/ceph-public.d52-54-77-77-01-01.mon.keyring: (2) No such file or directory
2018-10-04 12:24:53.035590 7f0168094f40 -1 ceph-mon: error creating monfs: (2) No such file or directory
---- End output of chown ceph:ceph /var/chef/cache/ceph-public.d52-54-77-77-01-01.mon.keyring ; ceph-mon --mkfs -i public.d52-54-77-77-01-01 --keyring '/var/chef/cache/ceph-public.d52-54-77-77-01-01.mon.keyring' --setuser ceph --setgroup ceph ----
Ran chown ceph:ceph /var/chef/cache/ceph-public.d52-54-77-77-01-01.mon.keyring ; ceph-mon --mkfs -i public.d52-54-77-77-01-01 --keyring '/var/chef/cache/ceph-public.d52-54-77-77-01-01.mon.keyring' --setuser ceph --setgroup ceph returned 1



Resource Declaration:
---------------------
# In /var/chef/cache/cookbooks/ceph/recipes/mon.rb


109:   execute "ceph-mon mkfs" do
110:     command "chown ceph:ceph #{keyring} ; ceph-mon --mkfs -i #{mon_name} " \
111:             "--keyring '#{keyring}' --setuser ceph --setgroup ceph"
112:     action :nothing
113:   end
114: 


Compiled Resource:
------------------
# Declared in /var/chef/cache/cookbooks/ceph/recipes/mon.rb:109:in `from_file'


execute("ceph-mon mkfs") do
  action [:nothing]
  retries 0
  retry_delay 2
  command "chown ceph:ceph /var/chef/cache/ceph-public.d52-54-77-77-01-01.mon.keyring ; ceph-mon --mkfs -i public.d52-54-77-77-01-01 --keyring '/var/chef/cache/ceph-public.d52-54-77-77-01-01.mon.keyring' --setuser ceph --setgroup ceph"
  backup 5
  returns 0
  cookbook_name "ceph"
  recipe_name "mon"
end


[2018-10-04T12:24:53+00:00] ERROR: Running exception handlers
[2018-10-04T12:24:53+00:00] FATAL: Saving node information to /var/chef/cache/failed-run-data.json
[2018-10-04T12:24:53+00:00] ERROR: Exception handlers complete
[2018-10-04T12:24:53+00:00] FATAL: Stacktrace dumped to /var/chef/cache/chef-stacktrace.out
[2018-10-04T12:24:53+00:00] FATAL: Mixlib::ShellOut::ShellCommandFailed: execute[ceph-mon mkfs] (ceph::mon line 109) had an error: Mixlib::ShellOut::ShellCommandFailed: Expected process to exit with [0], but received '1'
---- Begin output of chown ceph:ceph /var/chef/cache/ceph-public.d52-54-77-77-01-01.mon.keyring ; ceph-mon --mkfs -i public.d52-54-77-77-01-01 --keyring '/var/chef/cache/ceph-public.d52-54-77-77-01-01.mon.keyring' --setuser ceph --setgroup ceph ----
STDOUT: 
STDERR: chown: cannot access '/var/chef/cache/ceph-public.d52-54-77-77-01-01.mon.keyring': No such file or directory
2018-10-04 12:24:53.035565 7f0168094f40 -1 mon.public.d52-54-77-77-01-01@-1(probing) e0 unable to find a keyring file on /var/chef/cache/ceph-public.d52-54-77-77-01-01.mon.keyring: (2) No such file or directory
2018-10-04 12:24:53.035590 7f0168094f40 -1 ceph-mon: error creating monfs: (2) No such file or directory
---- End output of chown ceph:ceph /var/chef/cache/ceph-public.d52-54-77-77-01-01.mon.keyring ; ceph-mon --mkfs -i public.d52-54-77-77-01-01 --keyring '/var/chef/cache/ceph-public.d52-54-77-77-01-01.mon.keyring' --setuser ceph --setgroup ceph ----
Ran chown ceph:ceph /var/chef/cache/ceph-public.d52-54-77-77-01-01.mon.keyring ; ceph-mon --mkfs -i public.d52-54-77-77-01-01 --keyring '/var/chef/cache/ceph-public.d52-54-77-77-01-01.mon.keyring' --setuser ceph --setgroup ceph returned 1
```

tserong commented:
OK, that's weird. You didn't change anything related to mon keyring creation AFAICT, so why are those keyring files inaccessible?

bmwiedemann commented:
maybe the added package lines did something unexpected or something between SES4 and 5 packaging changed.

tserong commented:
I don't think this failure can be due to packaging changes -- that keyring file is created by the ceph barclamp during initial mon creation:

https://github.com/crowbar/crowbar-ceph/blob/0a96245e4bff633483c12a417e733402ee99d407/chef/cookbooks/ceph/recipes/mon.rb#L46-L122

It should have been created by one of  the three `ceph-authtool` invocations. Is there any indication earlier in the logs of ceph-authtool failing somehow?

jsuchome commented:
Hi @bmwiedemann, any update on this one?

bmwiedemann commented:
No, I was busy working on other parts and dont know how to continue here. Feel free to play around and improve this PR.

bmwiedemann commented:
ceph playground is open again: http://crowbar.vi16.cloud.suse.de/crowbar/ceph/1.0/proposals/default
